{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Copyright 2020 by Luke Selberg, Solis-Lemus Lab, WID.\n",
    "# All rights reserved.\n",
    "# This file is part of the BioKlustering Website.\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.cluster import MeanShift\n",
    "from sklearn import preprocessing\n",
    "import numpy as np\n",
    "import os\n",
    "from helpers import plotly_dash_show_plot\n",
    "\n",
    "def parseFasta(data):\n",
    "    # print(data)\n",
    "    d = {fasta.id : str(fasta.seq) for fasta in SeqIO.parse(data, \"fasta\")}\n",
    "    pd.DataFrame([d])\n",
    "\n",
    "    s = pd.Series(d, name='Sequence')\n",
    "    s.index.name = 'ID'\n",
    "    s.reset_index()\n",
    "    return pd.DataFrame(s)\n",
    "\n",
    "def kmerXTable(s, a, b):\n",
    "    tfid_vector = TfidfVectorizer(analyzer='char', ngram_range=(a,b))\n",
    "    s_hat = tfid_vector.fit_transform(s.Sequence)\n",
    "    kmerNames = tfid_vector.get_feature_names()\n",
    "    kmers = s_hat.toarray()\n",
    "    return pd.DataFrame(kmers,columns=kmerNames, index = s.index)\n",
    "\n",
    "# credit to chunrong\n",
    "def read_fasta_sequences(sequence_paths):\n",
    "    all_sequences = pd.DataFrame()\n",
    "    sequence = parseFasta(sequence_paths)\n",
    "    all_sequences = pd.concat([all_sequences, sequence])\n",
    "    # for path in sequence_paths:\n",
    "        # path = os.path.join(\"media\", path)\n",
    "        # sequence = parseFasta(path)\n",
    "        # all_sequences = pd.concat([all_sequences, sequence])\n",
    "    return all_sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def get_unique_numbers(numbers):\n",
    "\n",
    "    list_of_unique_numbers = []\n",
    "\n",
    "    unique_numbers = set(numbers)\n",
    "\n",
    "    for number in unique_numbers:\n",
    "        list_of_unique_numbers.append(number)\n",
    "\n",
    "    return list_of_unique_numbers\n",
    "\n",
    "def kmeans_semiSupervised(userId, fasta, klength_min, klength_max, rNum, y_hat, method):\n",
    "    inputData = read_fasta_sequences(fasta)\n",
    "    inputData[\"Sequence\"] = inputData[\"Sequence\"].apply(lambda x: x.replace(\"-\", \"\"))\n",
    "    kmerXTableInput = kmerXTable(inputData, klength_min, klength_max)\n",
    "\n",
    "    PCAembedding = PCA(n_components=10)\n",
    "    NkmerXTableInput = preprocessing.normalize(kmerXTableInput)\n",
    "    PCAembedding_low = PCAembedding.fit_transform(NkmerXTableInput)\n",
    "\n",
    "    ms = MeanShift()\n",
    "    ms.fit(PCAembedding_low)\n",
    "    cluster_centers = ms.cluster_centers_\n",
    "\n",
    "    import warnings\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        kmms = KMeans(init = cluster_centers, n_clusters = len(cluster_centers))\n",
    "        kmms_labels = kmms.fit_predict(PCAembedding_low)\n",
    "\n",
    "    kmerXTableInput[\"pLabels\"] = kmms_labels\n",
    "    kmerXTableInput[\"aLabels\"] = actual_labels = y_hat.tolist()\n",
    "    newLabelsClusters = dict()\n",
    "    unique_actual_labels = get_unique_numbers(kmerXTableInput[\"aLabels\"])\n",
    "    for actual_label in unique_actual_labels:\n",
    "        newLabelsClusters[actual_label] = kmerXTableInput[kmerXTableInput[\"aLabels\"] == actual_label][\"pLabels\"].tolist()\n",
    "\n",
    "    print(newLabelsClusters)\n",
    "    unique_predicted_labels = get_unique_numbers(kmms_labels)\n",
    "    new_labels_dict = dict()\n",
    "    for plabel in unique_predicted_labels:\n",
    "        l = {}\n",
    "        for key in newLabelsClusters.keys():\n",
    "            if key != -1:\n",
    "                l[key] = newLabelsClusters[key].count(plabel)\n",
    "        new_labels_dict[plabel] = max(l, key=l.get)\n",
    "\n",
    "    newLabels = []\n",
    "    print(f\"new_labels_dict: {new_labels_dict}\")\n",
    "    for i in range(len(kmms_labels)):\n",
    "        if actual_labels[i] == -1:\n",
    "            newLabels.append(new_labels_dict[kmms_labels[i]])\n",
    "        else:\n",
    "            newLabels.append(actual_labels[i])\n",
    "\n",
    "    print(f\"new labels: {newLabels}\")\n",
    "    kmerTable = kmerXTableInput.drop(columns=[\"pLabels\", \"aLabels\"])\n",
    "    plotly_kmertable = kmerTable\n",
    "    plotly_labels = np.array(newLabels)\n",
    "    if method == \"PCA\":\n",
    "        plotly_kmertable = preprocessing.normalize(kmerTable)\n",
    "    plotly_div = plotly_dash_show_plot(userId, plotly_kmertable, plotly_labels, \"Semi-supervised Kmeans\", method)\n",
    "\n",
    "    inputData.insert(0, \"Labels\", newLabels)\n",
    "\n",
    "    return [[inputData], [plotly_div]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{66.0: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0], 20.0: [0], 52.0: [0], 55.0: [0], 57.0: [0], 59.0: [1, 0, 3, 1, 1, 1, 1, 2, 3, 1, 1, 0, 1, 1, 1, 1], -1.0: [1, 1, 3, 1, 1, 2, 2, 1, 1, 1, 2, 1, 1, 1, 1, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0]}\n",
      "new_labels_dict: {0: 66.0, 1: 59.0, 2: 59.0, 3: 59.0}\n",
      "new labels: [59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 59.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 59.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 66.0, 20.0, 55.0, 66.0, 66.0, 52.0, 66.0, 66.0, 57.0, 66.0]\n"
     ]
    }
   ],
   "source": [
    "from numpy import genfromtxt\n",
    "#kmeans_semiSupervised(userId, fasta, klength_min, klength_max, rNum, y_hat, method)\n",
    "PATH01 = \"sequence_59.fasta\"\n",
    "known_labels = genfromtxt('labels_59.csv', delimiter=',')\n",
    "data = PATH01\n",
    "labels_true = [-1 for x in range(100)]\n",
    "result = kmeans_semiSupervised(userId = 1, fasta = PATH01, klength_min = 6, klength_max = 7, rNum = 0, y_hat = known_labels, method=\"PCA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def kmeans(userId, fasta, klength_min, klength_max, rNum, cNum, method):\n",
    "    inputData = read_fasta_sequences(fasta)\n",
    "    inputData[\"Sequence\"] = inputData[\"Sequence\"].apply(lambda x: x.replace(\"-\", \"\"))\n",
    "\n",
    "    kmerXTableInput = kmerXTable(inputData, klength_min, klength_max)\n",
    "    km = KMeans(random_state = rNum, n_clusters = cNum)\n",
    "    km.fit(kmerXTableInput)\n",
    "    y_hat = km.predict(kmerXTableInput)\n",
    "    unique_predict_labels = get_unique_numbers(y_hat)\n",
    "    plotly_kmertable = kmerXTableInput\n",
    "    if method == \"PCA\":\n",
    "        plotly_kmertable = preprocessing.normalize(kmerXTableInput)\n",
    "    plot_div = plotly_dash_show_plot(userId, plotly_kmertable, y_hat, \"Unsupervised Kmeans\", method)\n",
    "    inputData.insert(0, \"Labels\", y_hat)\n",
    "\n",
    "    return [[inputData], [plot_div]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence_59.fasta\n",
      "plotly_kmertable:               aaaaaa   aaaaaaa   aaaaaag   aaaaaat    aaaaac   aaaaaca  \\\n",
      "ID                                                                       \n",
      "OA979645.1  0.002543  0.000000  0.001297  0.001271  0.020343  0.010171   \n",
      "OD912801.1  0.002540  0.000000  0.001296  0.001270  0.020324  0.010162   \n",
      "FR998715.1  0.002755  0.000000  0.001405  0.001378  0.022041  0.011021   \n",
      "OA966210.1  0.001861  0.000000  0.000949  0.000931  0.014891  0.007445   \n",
      "OA979303.1  0.002540  0.000000  0.001296  0.001270  0.020322  0.010161   \n",
      "...              ...       ...       ...       ...       ...       ...   \n",
      "LR883375.1  0.001378  0.000000  0.000000  0.001378  0.020669  0.011023   \n",
      "MT470109.1  0.040994  0.098533  0.001394  0.001366  0.021863  0.010932   \n",
      "MZ332124.1  0.002758  0.000000  0.001407  0.001379  0.022063  0.011031   \n",
      "MT956916.1  0.041009  0.098569  0.001394  0.001367  0.021871  0.010936   \n",
      "LC632047.1  0.041002  0.098553  0.001394  0.001367  0.021868  0.010934   \n",
      "\n",
      "             aaaaacc   aaaaacg   aaaaact    aaaaag  ...  ytatta  ytattaa  \\\n",
      "ID                                                  ...                    \n",
      "OA979645.1  0.001284  0.001271  0.007629  0.027971  ...     0.0      0.0   \n",
      "OD912801.1  0.001283  0.001270  0.007621  0.026675  ...     0.0      0.0   \n",
      "FR998715.1  0.001391  0.001378  0.008266  0.030307  ...     0.0      0.0   \n",
      "OA966210.1  0.000940  0.000931  0.005584  0.020475  ...     0.0      0.0   \n",
      "OA979303.1  0.001283  0.001270  0.007621  0.027943  ...     0.0      0.0   \n",
      "...              ...       ...       ...       ...  ...     ...      ...   \n",
      "LR883375.1  0.000000  0.001378  0.008268  0.023425  ...     0.0      0.0   \n",
      "MT470109.1  0.001380  0.001366  0.008199  0.030062  ...     0.0      0.0   \n",
      "MZ332124.1  0.001393  0.001379  0.008274  0.030336  ...     0.0      0.0   \n",
      "MT956916.1  0.001381  0.001367  0.008202  0.030073  ...     0.0      0.0   \n",
      "LC632047.1  0.001380  0.001367  0.008200  0.030068  ...     0.0      0.0   \n",
      "\n",
      "            ytggtg  ytggtgt  ytggtt  ytggttt  yttgaa  yttgaaa  yttgta  yttgtac  \n",
      "ID                                                                              \n",
      "OA979645.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "OD912801.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "FR998715.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "OA966210.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "OA979303.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "...            ...      ...     ...      ...     ...      ...     ...      ...  \n",
      "LR883375.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "MT470109.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "MZ332124.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "MT956916.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "LC632047.1     0.0      0.0     0.0      0.0     0.0      0.0     0.0      0.0  \n",
      "\n",
      "[100 rows x 16665 columns]\n",
      "y_hat: [0 0 1 2 0 2 0 0 0 0 0 3 0 3 3 0 2 0 0 0 3 0 0 1 0 0 0 0 0 0 2 3 0 1 1 1 1\n",
      " 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1\n",
      " 1 1 1 1 0 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1]\n"
     ]
    }
   ],
   "source": [
    "result = kmeans(userId = 1, fasta = PATH01, klength_min = 6, klength_max = 7, cNum = 4, rNum = 0, method=\"PCA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0     59\n",
      "1     -1\n",
      "2     59\n",
      "3     59\n",
      "4     -1\n",
      "      ..\n",
      "95    52\n",
      "96    -1\n",
      "97    -1\n",
      "98    57\n",
      "99    -1\n",
      "Name: 0, Length: 100, dtype: int64\n",
      "<class 'numpy.ndarray'>\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'PredictInfo' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-de0a347f411a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    197\u001b[0m \u001b[0mPATH01\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"sequence_59.fasta\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m \u001b[0;31m# def get_predictions_semi(userId, path, k_min, k_max, num_class, cov_type, seed, labels, method)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 199\u001b[0;31m [labels], [plot] = get_predictions_semi(userId=0, path = PATH01, k_min = k_min, k_max = k_max, num_class = num_class,\n\u001b[0m\u001b[1;32m    200\u001b[0m                                         cov_type = cov_type, seed = seed, labels = labels_50, method = 'PCA')\n\u001b[1;32m    201\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-12-de0a347f411a>\u001b[0m in \u001b[0;36mget_predictions_semi\u001b[0;34m(userId, path, k_min, k_max, num_class, cov_type, seed, labels, method)\u001b[0m\n\u001b[1;32m    119\u001b[0m     \u001b[0;31m# update parameters for predictInfo object (i.e. for front end)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    120\u001b[0m     \u001b[0macc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcal_accuracy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpredictions_final\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 121\u001b[0;31m     \u001b[0mupdate_parameters\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muserId\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"accuracy\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0macc\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    122\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0moutput_df\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mplot_div\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Desktop/CS/Lab/bioklustering/BioKlustering-Website/mlmodel/parser/helpers.py\u001b[0m in \u001b[0;36mupdate_parameters\u001b[0;34m(userId, new_params)\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[0;31m# new_params: a dictinary where key is the param name and value is the param value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     79\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mupdate_parameters\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muserId\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_params\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 80\u001b[0;31m     \u001b[0mpredict_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPredictInfo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobjects\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muser\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0muserId\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     81\u001b[0m     \u001b[0mpredict_info_params\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpredict_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"parameters\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     82\u001b[0m     \u001b[0mpredict_info_params_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpredict_info_params\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'PredictInfo' is not defined"
     ]
    }
   ],
   "source": [
    "# import packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO, AlignIO\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.mixture import GaussianMixture as GMM\n",
    "import os\n",
    "\n",
    "from helpers import *\n",
    "\n",
    "\n",
    "def parseFasta(data):\n",
    "    d = {fasta.id: str(fasta.seq) for fasta in SeqIO.parse(data, \"fasta\")}\n",
    "    pd.DataFrame([d])\n",
    "    s = pd.Series(d, name='Sequence')\n",
    "    s.index.name = 'ID'\n",
    "    s.reset_index()\n",
    "    return pd.DataFrame(s)\n",
    "\n",
    "\n",
    "def get_kmer_table(path, k_min, k_max):\n",
    "    genes, gene_len, output_df = read_fasta(path)\n",
    "    count_vect = CountVectorizer(analyzer='char', ngram_range=(k_min, k_max))\n",
    "    X = count_vect.fit_transform(genes)\n",
    "    chars = count_vect.get_feature_names()\n",
    "    kmers = X.toarray()\n",
    "    kmer_freq = []\n",
    "    for i in range(len(genes)):\n",
    "        kmer_freq = pd.concat([kmer_freq, kmers[i] / gene_len[i]],ignore_index=True) #kmer_freq.append(kmers[i] / gene_len[i])\n",
    "    input = pd.DataFrame(kmer_freq, columns=chars)\n",
    "    return input, output_df\n",
    "\n",
    "\n",
    "def get_gene_sequences(filename):\n",
    "    genes = []\n",
    "    for record in SeqIO.parse(filename, \"fasta\"):\n",
    "        genes.append(str(record.seq))\n",
    "    return genes\n",
    "\n",
    "\n",
    "# genes: a list of gene sequences, which can directly be generated from get_gene_sequences().\n",
    "def get_gene_len(genes):\n",
    "    gene_len = []\n",
    "\n",
    "    for i in range(len(genes)):\n",
    "        gene_len.append(len(genes[i]))\n",
    "    return gene_len\n",
    "\n",
    "\n",
    "def read_fasta(paths):\n",
    "    all_genes = []\n",
    "    all_gene_len = []\n",
    "    output_df = pd.DataFrame()\n",
    "    temp = paths\n",
    "    paths = [temp]\n",
    "    # print(paths)\n",
    "    for path in paths:\n",
    "        # path = os.path.join('media', path)\n",
    "        virus = parseFasta(path)\n",
    "        output_df = pd.concat([output_df, virus])\n",
    "        virus = virus.drop_duplicates(keep=\"last\")\n",
    "        genes = list(virus['Sequence'])\n",
    "        genes_seq = get_gene_sequences(path)\n",
    "        gene_len = get_gene_len(genes_seq)\n",
    "        all_genes = all_genes + genes_seq\n",
    "        all_gene_len = all_gene_len + gene_len\n",
    "    return all_genes, all_gene_len, output_df\n",
    "\n",
    "\n",
    "def get_predictions(userId, path, k_min, k_max, num_class, cov_type, seed, method):\n",
    "    kmer_table, output_df = get_kmer_table(path, k_min, k_max)\n",
    "    gmm = GMM(n_components=num_class, covariance_type=cov_type, random_state=seed).fit(kmer_table)\n",
    "    predictions = gmm.predict(kmer_table)\n",
    "    plot_div = plotly_dash_show_plot(userId, kmer_table, predictions, \"Unsupervised Gaussian Mixture Model\", method)\n",
    "    output_df.insert(0, \"Labels\", predictions)\n",
    "    return [[output_df], [plot_div]]\n",
    "\n",
    "\n",
    "# modified for website\n",
    "def get_predictions_semi(userId, path, k_min, k_max, num_class, cov_type, seed, labels, method):\n",
    "    targets = []\n",
    "    kmer_table, output_df = get_kmer_table(path, k_min, k_max)\n",
    "\n",
    "    finalDf = pd.concat([kmer_table, labels], axis=1)\n",
    "    gmm = GMM(n_components=num_class, covariance_type=cov_type, random_state=seed)\n",
    "    print(finalDf[0])\n",
    "    for i in range(num_class):\n",
    "        if i in list(finalDf[0]):\n",
    "            targets = pd.concat([targets, i],ignore_index=True)\n",
    "            #targets.append(i)\n",
    "    if len(targets) == num_class:\n",
    "        gmm.means_init = np.array([kmer_table[finalDf[0] == i].mean(axis=0) for i in targets])\n",
    "    gmm.fit(kmer_table)\n",
    "    predictions = gmm.predict(kmer_table)\n",
    "    print(type(predictions))\n",
    "    unique_given_labels = get_unique_numbers(labels)\n",
    "    given_labels_count = {}\n",
    "    labels_list = list(labels)\n",
    "    for label in unique_given_labels:\n",
    "        given_labels_count[label] = labels_list.count(label)\n",
    "    unique_predicted_labels = get_unique_numbers(predictions)\n",
    "    predicted_labels_count = {}\n",
    "    for label in unique_predicted_labels:\n",
    "        predicted_labels_count[label] = (predictions == label).sum()\n",
    "    del given_labels_count[-1]\n",
    "    given_labels_count = sorted(given_labels_count.items(), key=lambda x: x[1], reverse=True)\n",
    "    predicted_labels_count = sorted(predicted_labels_count.items(), key=lambda x: x[1], reverse=True)\n",
    "    map_predict_to_actual = {}\n",
    "    for i in range(len(predicted_labels_count)):\n",
    "        map_predict_to_actual[predicted_labels_count[i][0]] = given_labels_count[i][0]\n",
    "    predictions_final = []\n",
    "    for i in range(len(predictions)):\n",
    "        if labels[i] == -1:\n",
    "            predictions_final.append(map_predict_to_actual[predictions[i]])\n",
    "        else:\n",
    "            predictions_final.append(labels[i])\n",
    "    predictions = np.array(predictions_final)\n",
    "    plot_div = plotly_dash_show_plot(userId, kmer_table, predictions, \"Semi-supervised Gaussian Mixture Model\", method)\n",
    "    output_df.insert(0, \"Labels\", predictions)\n",
    "    # update parameters for predictInfo object (i.e. for front end)\n",
    "    acc = cal_accuracy(labels, predictions_final)\n",
    "    update_parameters(userId, {\"accuracy\": acc})\n",
    "    return [[output_df], [plot_div]]\n",
    "\n",
    "\n",
    "def get_unique_numbers(numbers):\n",
    "\n",
    "    list_of_unique_numbers = []\n",
    "\n",
    "    unique_numbers = set(numbers)\n",
    "\n",
    "    for number in unique_numbers:\n",
    "        list_of_unique_numbers.append(number)\n",
    "\n",
    "    return list_of_unique_numbers\n",
    "\n",
    "def cal_accuracy(labels, predictions):\n",
    "    err = 0\n",
    "    total_len = len(labels)\n",
    "    for i in range(len(labels)):\n",
    "        if labels[i] == -1:\n",
    "            total_len = total_len - 1\n",
    "            continue\n",
    "        if labels[i] != predictions[i]:\n",
    "            err += 1\n",
    "\n",
    "    return 1 - err / total_len\n",
    "\n",
    "\n",
    "def get_predictions_semi_original(path, k_min, k_max, num_class, cov_type, seed, labels):\n",
    "    kmer_table, output_df = get_kmer_table(path, k_min, k_max)\n",
    "    finalDf = pd.concat([kmer_table, pd.Series(labels)], axis=1)\n",
    "    gmm = GMM(n_components=num_class, covariance_type=cov_type, random_state=seed)\n",
    "    gmm.means_init = np.array([kmer_table[finalDf.Labels == i].mean(axis=0) for i in range(num_class)])\n",
    "    gmm.fit(kmer_table)\n",
    "    predictions = gmm.predict(kmer_table)\n",
    "    return predictions\n",
    "\n",
    "\n",
    "def model_selection(userId, path, labels, num_class, seed, method):\n",
    "    best_accu = 0\n",
    "    best_prediction = []\n",
    "    cov_type = ['full', 'diag', 'tied', 'spherical']\n",
    "    k_min = [2, 3, 4]\n",
    "    k_max = [2, 3, 4, 5]\n",
    "    for cov in cov_type:\n",
    "        for k1 in k_min:\n",
    "            for k2 in k_max:\n",
    "                if k2 >= k1:\n",
    "                    prediction = get_predictions_semi_original(path, k1, k2, num_class, cov, 0, labels)\n",
    "                    accu = cal_accuracy(labels, prediction)\n",
    "                    if accu > best_accu:\n",
    "                        best_accu = accu\n",
    "                        best_kmin = k1\n",
    "                        best_kmax = k2\n",
    "                        best_cov = cov\n",
    "                        best_prediction = prediction\n",
    "\n",
    "    # update parameters for predictInfo object (i.e. for front end)\n",
    "    new_params = {\n",
    "        'k_min': best_kmin,\n",
    "        'k_max': best_kmax,\n",
    "        'cov_type': best_cov,\n",
    "        'accuracy': best_accu\n",
    "    }\n",
    "    update_parameters(userId, new_params)\n",
    "    return get_predictions_semi(userId, path, best_kmin, best_kmax, num_class, best_cov, seed, labels, method)\n",
    "    # return best_kmin,best_kmax,best_cov,best_prediction\n",
    "\n",
    "\n",
    "k_min = 2\n",
    "k_max = 3\n",
    "num_class = 2\n",
    "cov_type = 'full'\n",
    "seed = 0\n",
    "labels_50 = pd.read_csv('labels_59.csv', delimiter=',', header=None)\n",
    "labels_50 = pd.Series(labels_50[0])\n",
    "PATH01 = \"sequence_59.fasta\"\n",
    "# def get_predictions_semi(userId, path, k_min, k_max, num_class, cov_type, seed, labels, method)\n",
    "[labels], [plot] = get_predictions_semi(userId=0, path = PATH01, k_min = k_min, k_max = k_max, num_class = num_class,\n",
    "                                        cov_type = cov_type, seed = seed, labels = labels_50, method = 'PCA')\n",
    "print(labels)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
